### PR TITLE
137 vector change

### PR DESCRIPTION
**Pull request by rsantellan**
Remove unused functions.
Make function private.
Copy colors to all colors.2da
Update to use a ENUM and map to the colors in the table.
I create this because the old one has a lot of commit of a merge.

## Description
Externalize the colors of the message to table. #137

## Checklist

- [X] Commit messages are descriptive and explain the rationale for changes
- [X] I used the same coding style as the surrounding code 
- [X] I have tested the proposed changes
- [X] I extended the documentation, if necessary
- [ ] The proposed change builds also on our build bots (check after submission)

### CODE REVIEW

File: gemrb/core/DisplayMessage.cpp [hunk 1]
```diff
@@ -144,7 +146,7 @@ Color DisplayMessage::GetSpeakerColor(String& name, const Scriptable *&speaker)
 		case ST_TRIGGER:
 		case ST_PROXIMITY:
 		case ST_TRAVEL:
-			name = core->GetString( speaker->DialogName );
+			name = core->GetString(speaker->DialogName );
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.cpp:149]: this is an unrelated change, so it shouldn't be here.

File: gemrb/core/DisplayMessage.cpp [hunk 2]
```diff
@@ -187,30 +190,65 @@ void DisplayMessage::DisplayString(String text, const Color &color, Scriptable *
 	}
 }
 
+std::vector<Color> DisplayMessage::GetAllColors() const
+{
+	std::vector<Color> auxiliaryColors;
+	if (auxiliaryColors.empty()) {
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.cpp:196]: this is always true

File: gemrb/core/DisplayMessage.cpp [hunk 3]
```diff
@@ -187,30 +190,65 @@ void DisplayMessage::DisplayString(String text, const Color &color, Scriptable *
...
+		AutoTable colorTable = gamedata->LoadTable("colors", true);
+		assert(colorTable);
+		for (TableMgr::index_t r = 0; r < colorTable->GetRowCount(); r++) {
+			ieDword c = colorTable->QueryFieldUnsigned<ieDword>(r, 0);
+			auxiliaryColors.push_back(Color(c));
+		}
+	}
+	return auxiliaryColors;
+}
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.cpp:205]: I disagree with this approach; why store all the colors again? We have them in GameData already. Like I explained before, we just need their names. And then also the new `DisplayMessage::GetColor` can be removed.

File: gemrb/core/DisplayMessage.h [hunk 4]
```diff
@@ -33,16 +33,20 @@
...
+/** GUIColors options */
+enum class GUIColors {
+	DIALOG = 12,
+	DIALOGPARTY = 13,
+	GOLD = 11,
+	RED = 8,
+	XPCHANGE = 10,
+	LIGHTGREY = 9,
+	WHITE = 7
+};
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.h:49]: please order this and add a comment why it doesn't start at 0

File: gemrb/core/DisplayMessage.cpp (hunk 3 above)
Comment (rsantellan) [reply gemrb/core/DisplayMessage.cpp:205]: The problem is that on https://github.com/gemrb/gemrb/blob/5834c9ada3b08f2c77604351f20e8a82b4b76bed/gemrb/core/GameData.cpp#L640 the colors created is ` std::map<std::string, Color> colors;` so to map my enum I pass it as vector Color. 
The problem with this approach is that I have 13 colors defined on colors.2da so I have to map the positions to the enum.
I don't like this at all, for me will be better to map the enum to the names so there is no mistake there.

File: gemrb/core/DisplayMessage.h [hunk 5]
```diff
@@ -37,7 +37,7 @@
 
 namespace GemRB {
 
-/** GUIColors options */
+/** GUIColors options. This starts at 7 because is the row in the colors.2da that corresponds to white color.*/
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.h:40]: A better comment would be that the first six colors from colors.2da are not indexed here because they are only used on the python side.

File: gemrb/core/DisplayMessage.h [hunk 6]
```diff
@@ -39,13 +39,13 @@ namespace GemRB {
...
-	GOLD = 11,
+	WHITE = 7,
 	RED = 8,
-	XPCHANGE = 10,
 	LIGHTGREY = 9,
-	WHITE = 7
+	XPCHANGE = 10,
+	GOLD = 11,
+	DIALOG = 12,
+	DIALOGPARTY = 13
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.h:48]: Now that they're in order, you don't need to manually assign values to all of them. WHITE, as the first, is enough.

File: gemrb/core/DisplayMessage.cpp [hunk 7]
```diff
@@ -190,28 +188,26 @@ void DisplayMessage::DisplayString(String text, const Color &color, Scriptable *
...
-	if (size < colortInt) {
-		return ColorRed;
+	
+	int colorInt = static_cast<int>(color) - 1;
+	if (static_cast<int>(GUIColorNames.size()) < colorInt) {
+		// It never should happend this, but just for been defensive.
+		colorInt = 0;
 	}
-	return colors.at(colortInt);
+	return gamedata->GetColor(GUIColorNames[colorInt]);
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.cpp:210]: Let's use `std::map<GUIColors, std::string>` then for GUIColorNames, so this function can be simplified or removed altogether. Because it's a map, it also means you don't have to check for non-existing keys, as new entries would get created automatically and eventually the default color from the table would be used.

File: gemrb/core/DisplayMessage.cpp [hunk 8]
```diff
@@ -187,30 +188,63 @@ void DisplayMessage::DisplayString(String text, const Color &color, Scriptable *
 	}
 }
 
+std::vector<std::string> DisplayMessage::GetAllColors() const
+{
+	std::vector<std::string> auxiliaryColors;
+	AutoTable colorTable = gamedata->LoadTable("colors", true);
+	assert(colorTable);
+	for (TableMgr::index_t r = 0; r < colorTable->GetRowCount(); r++) {
```
Comment (miscco) [inline gemrb/core/DisplayMessage.cpp:196]: This is missing a call to `reserve` which will lead to a whole bunch of allocations
```suggestion
        auxiliaryColors.reserve(colorTable->GetRowCount());
	for (TableMgr::index_t r = 0; r < colorTable->GetRowCount(); r++) {
```

File: gemrb/core/DisplayMessage.cpp [hunk 9]
```diff
@@ -187,30 +188,63 @@ void DisplayMessage::DisplayString(String text, const Color &color, Scriptable *
...
+		auxiliaryColors.push_back(colorTable->GetRowName(r));
+	}
+	return auxiliaryColors;
+}
+
+Color DisplayMessage::GetColor(GUIColors color) const
+{
+
```
Comment (miscco) [inline gemrb/core/DisplayMessage.cpp:204]:
```suggestion
```

File: gemrb/core/DisplayMessage.cpp [hunk 10]
```diff
@@ -187,30 +188,63 @@ void DisplayMessage::DisplayString(String text, const Color &color, Scriptable *
 	}
 }
 
+std::vector<std::string> DisplayMessage::GetAllColors() const
+{
+	std::vector<std::string> auxiliaryColors;
```
Comment (miscco) [inline gemrb/core/DisplayMessage.cpp:193]: Arent those names static? A view type like `std::string_view` would be much better here

Comment (lynxlynxlynx) [reply gemrb/core/DisplayMessage.cpp:193]: I don't see any benefit, but there's no reason for the function to work with a temporary either.

File: gemrb/core/DisplayMessage.cpp (hunk 7 above)
Comment (rsantellan) [reply gemrb/core/DisplayMessage.cpp:210]: @lynxlynxlynx this will not be the same as the function **gamedata->GetColor** at the end ?

Comment (lynxlynxlynx) [reply gemrb/core/DisplayMessage.cpp:211]: No, the gamedata version will still be the main one to get the precached Color values by color (row) name.
When I say that this function here can probably go away, I'm thinking of just replacing calls to it by `gamedata->GetColor(GUIColorNames[color])` (color being one of the GUIColors values).

Comment (rsantellan) [reply gemrb/core/DisplayMessage.cpp:211]: Sorry I don't understand this. Now I have:

- A **std::vector<std::string>** that has all the string names on the **colors.2da table.**
- A **Enum** that has the row where the color is set.
- Using the **Enum** I use that position on the vector to retrieve the color name to use on **gamedata->GetColor**

And the proposed change will work something like this:
If I use a **std::map<GUIColors, std::string>** I will have to also have a way to convert the **Enum** to a position on the string, so I have to keep the **std::vector<std::string>** and also use the map, and then I will not need  **gamedata->GetColor** because I have the color there, but if not in the map I will call **gamedata->GetColor** ?

This is what I'm thinking, I'm not fully comprehend where the change is going or how it will work.

Comment (lynxlynxlynx) [reply gemrb/core/DisplayMessage.cpp:211]: > I will have to also have a way to convert the Enum to a position on the string
What string? The enum still holds the row numbers and the std::map will map these enum values to the row names. Exactly like the vector, except that the indices are properly typed (enum vs general int), so there's no need to convert the index. Or care about missing indexes (keys).

File: gemrb/core/DisplayMessage.cpp [hunk 11]
```diff
@@ -188,27 +188,24 @@ void DisplayMessage::DisplayString(String text, const Color &color, Scriptable *
 	}
 }
 
-std::vector<std::string> DisplayMessage::GetAllColors() const
+std::map<int, std::string> DisplayMessage::GetAllColors() const
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.cpp:191]: Why not the map I proposed? There's not only no need for the full range of integer keys, but more bugs are possible that way.

**Comment by rsantellan** [comment]
Closed in favor of  #1625
